File: 146-lru-cache/146-lru-cache.java
class LRUCache {
    LinkedListNode head = null;
    LinkedListNode tail = null;
    Map<Integer, LinkedListNode> map = new HashMap<>();
    int size;
    
    public LRUCache(int capacity) {
        size = capacity;
        
        // create dummy
        head = new LinkedListNode(-1, -1);
        tail = head;
    }
    
    public int get(int key) {
        LinkedListNode lln = map.get(key);
        if (lln == null) return -1;
        
        renewNode(lln);
        
        return lln.value;
    }
    
    public void put(int key, int value) {
        LinkedListNode lln = map.get(key);
        
        if (lln == null) {
            lln = new LinkedListNode(key, value);
            map.put(key, lln);
            appendNode(lln);
        } else {
            lln.value = value;
            renewNode(lln);
        }
        
        if (map.size() > size) {
            LinkedListNode remove = removeHead();
            if (remove != null) map.remove(remove.key);
        }
    }
    
    void renewNode(LinkedListNode lln) {
        if (lln == tail) return;
        
        lln.prev.next = lln.next;
        
        if (lln.next != null) lln.next.prev = lln.prev;
        
        appendNode(lln);
    }
    
    LinkedListNode removeHead() {
        if (head.next == null) return null;
        
        LinkedListNode remove = head.next;
        
        head.next = remove.next;
        if (remove.next.prev != null) remove.next.prev = head;
        
        return remove;
    }
    
    void appendNode(LinkedListNode lln) {
        if (tail == lln) return;
        
        tail.next = lln;
        lln.prev = tail;
        tail = lln;
        tail.next = null;
    }
    
    class LinkedListNode {
        public int key;
        public int value;
        
        public LinkedListNode prev = null;
        public LinkedListNode next = null;

        public LinkedListNode(int key, int value) {
            this.key = key;
            this.value = value;
        }
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: 146-lru-cache/146-lru-cache.scala
import scala.collection.mutable._

class LRUCache(_capacity: Int) {
    case class DL(val key: Int, var value: Int, var prev: DL = null, var next: DL = null)

    val head: DL = DL(-1, -1)
    val tail: DL = DL(-1, -1)
    head.prev = tail
    tail.next = head
    val map: Map[Int, DL] = Map()

    def get(key: Int): Int = {
        map.get(key)
            .map { node =>
                remove(node)
                putHead(node)
                node
            }
            .map(_.value)
            .getOrElse(-1)
    }

    def put(key: Int, value: Int):Unit = {
        if (map.contains(key)) {
            val node = map.get(key).get
            node.value = value
            remove(node)
            putHead(node)
        } else {
            val node = DL(key, value)
            map += (key -> node)
            putHead(node)

            if (map.size > _capacity) {
                val toBeRemoved = tail.next
                map.remove(toBeRemoved.key)
                remove(toBeRemoved)
            }
        }
    }

    private def remove(node: DL): Unit = {
        node.prev.next = node.next
        node.next.prev = node.prev
    }

    private def putHead(node: DL): Unit = {
        node.next = head
        node.prev = head.prev
        head.prev = node
        node.prev.next = node
    }

}


/**
 * Your LRUCache object will be instantiated and called as such:
 * var obj = new LRUCache(capacity)
 * var param_1 = obj.get(key)
 * obj.put(key,value)
 */